include/picongpu/fields/incidentField/Solver.hpp

```diff
@@ -0,0 +1,516 @@
+/* Copyright 2020-2021 Sergei Bastrakov
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+
+#include "picongpu/fields/Fields.hpp"
+#include "picongpu/fields/MaxwellSolver/Solvers.def"
+#include "picongpu/fields/absorber/Absorber.hpp"
+#include "picongpu/fields/incidentField/Profiles.hpp"
+#include "picongpu/fields/incidentField/Solver.kernel"
+#include "picongpu/fields/incidentField/Traits.hpp"
+
+#include <pmacc/math/Vector.hpp>
+
+#include <cstdint>
+#include <type_traits>
+
+
+namespace picongpu
+{
+    namespace fields
+    {
+        namespace incidentField
+        {
+            namespace detail
+            {
+                /** Internally used parameters of the incident field generation normal to the given axis
+                 *
+                 * @tparam T_axis boundary axis, 0 = x, 1 = y, 2 = z
+                 */
+                template<uint32_t T_axis>
+                struct Parameters
+                {
+                    /** Create a parameters instance
+                     *
+                     * @param cellDescription cell description for kernels
+                     */
+                    Parameters(MappingDesc const cellDescription) : cellDescription(cellDescription)
+                    {
+                    }
+
+                    /** Offset of the Huygens surface from min border of the global domain
+                     *
+                     * The offset is from the start of CORE + BORDER area.
+                     * Counted in full cells, the surface is additionally offset by 0.75 cells
+                     */
+                    pmacc::DataSpace<simDim> offsetMinBorder;
+
+                    /** Offset of the Huygens surface from max border of the global domain
+                     *
+                     * The offset is from the start of CORE + BORDER area.
+                     * Counted in full cells, the surface is additionally offset by 0.75 cells
+                     */
+                    pmacc::DataSpace<simDim> offsetMaxBorder;

```

Comment:
They can be combined, but with a doxygen group then. Which is less duplication, but more lines and I personally dislike it. But here it would probably fit.

Reply:
Okay - true. Thanks @sbastrakov for your explanation.